File: jw/week51/MinimumSizeSubarraySum.java
class Solution {
    public int minSubArrayLen(int target, int[] nums) {
        int size = nums.length;
        int s = 0;
        int e = 0;
        int sum = nums[0];
        int min = Integer.MAX_VALUE;
        
        while(s < size && e < size) {
            if(sum < target) {
                e+=1;
                if(e == size ) break;
                sum += nums[e];
            } else if(sum > target) {
                min = Math.min(e-s+1,min);
                if(e == s) {
                    e+=1;
                    if(e == size ) break;
                    sum += nums[e];
                } else {
                    s+=1;
                    sum -= nums[s-1];
                }
            } else {
                min = Math.min(e-s+1,min);
                e+= 1;
                if(e == size ) break;
                sum += nums[e];
            }
        }
        
        if(min == Integer.MAX_VALUE) min = 0;
        
        return min;
    }
}

File: sj/51week/Minimum Size Subarray Sum.js
/**
 * @param {number} target
 * @param {number[]} nums
 * @return {number}
 */
var minSubArrayLen = function(target, nums) {
    
    let left = 0;
 
    let sum =0;
    
    let res = Number.MAX_SAFE_INTEGER;
    
    for(let r = 0; r< nums.length; r++)
        {
            sum += nums[r];
            
            while(sum >= target)
                {
                    res = Math.min(res, r-left +1);
                    
                    sum -= nums[left];
                    left = left +1;
                }
        }
    return res === Number.MAX_SAFE_INTEGER ? 0 : res;
};

//베스트 솔직히 거의 똑같음
public int minSubArrayLen(int s, int[] a) {
  if (a == null || a.length == 0)
    return 0;
  
  int i = 0, j = 0, sum = 0, min = Integer.MAX_VALUE;
  
  while (j < a.length) {
    sum += a[j++];
    
    while (sum >= s) {
      min = Math.min(min, j - i);
      sum -= a[i++];
    }
  }
  
  return min == Integer.MAX_VALUE ? 0 : min;
}
